### ISSUE TITLE
files that depend on a dynamic class (say C) are not recompiled when C's private fields become public

### ISSUE THREAD

**Issue by symingz** [sbt/sbt#7589]
## steps

SBT version: 1.10.0
Scala version: 3.4.2
Project setup: sample SBT hello-world project with two files: Main.scala and Bar.scala

### Main.scala:
```
@main def hello(): Unit =
  val tt = Bar()
  println(tt.pv)
```
### Bar.scala
```
import scala.language.dynamics

class Bar extends Dynamic:
  private val pv = 31415
  def selectDynamic(name: String): Unit =
    println("selectDynamic: " + name)
```

1. SBT input
```
clean
run
```
output
```
selectDynamic: pv
() 
```
2. Change line `private val pv = 31415` in `Bar.scala` to `val pv = 31415`
3. SBT Input
```
run
```
output
```
selectDynamic: pv
()
```
4. SBT Input
```
clean
run
```
output
```
31415
```

## problem

Main.scala is not recompiled after Bar.scala is changed in Step 2.

## expectation

The output in Step 3 should be
```
31415
```
instead of
```
selectDynamic: pv
()
```
In other words, `clean` (as in Step 4) should not be needed to get the correct result.

**Comment by eed3si9n** [comment]
@symingz Thanks for the report. Does the similar code work for Scala 2.13?

**Comment by symingz** [comment]
No, Scala 2.13.12 has the same problem.

**Comment by eed3si9n** [comment]
Zinc tracks all functions calls by the names. My guess is that we do not treat dynamic calls as track-worthy names because there's normally nothing function that would match on the other side to pair, and thus when the new API showed up on the other side it didn't register as an invalidation.

https://github.com/sbt/zinc/blob/4eacff2a9bf5c8750bfc5096955065ce67f4e68a/internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala#L435-L436

**Comment by Friendseeker** [comment]
Can probably fix the issue by properly tracking name in API Phase as Eugene mentioned. Hopefully scalac emits some easy to work with AST to make this simple.

**Comment by Friendseeker** [comment]
For reference this is apparently a known issue 11 years ago...

https://github.com/gkossakowski/sbt/issues/4

It even has a pending test already written in https://github.com/sbt/zinc/blob/f644b6c22f9967ae3451031b686aac6ac4fac28a/internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesSpecification.scala#L213

**Comment by corevibe555** [comment]
@eed3si9n I would love to work on this.

**Comment by eed3si9n** [comment]
@corevibe555 Sure. The relevant change would be in sbt/zinc repository. As per usual, please check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) and:
1. Make sure you can reproduce the problem and the fix manually
2. Include tests for the change
3. Follow our AI assisted contributions guideline (please disclose AI usages in the PR description, if any)
4. Multiple contributors might work on this issue in parallel
5. Generally we recommend getting CI to pass on a forked repo before sending a pull request (pull requests that do not pass tests will be closed after a few days)
